scylla/src/transport/load_balancing/fallback.rs

```diff
@@ -31,3 +31,82 @@ impl<T: LoadBalancingPolicy> FallbackPolicy for T {
         Box::new(plan)
     }
 }
+
+/// Combines the primary plan and the fallback plan. Fallback is not generated until primary plan
+/// exhaustion.
+pub(crate) struct PlanWithFallback<'a, G: FnOnce(HashSet<SocketAddr>) -> Plan<'a>> {
+    source: PlanSource<'a>,
+    fallback_generator: Option<G>,
+}
+
+impl<'a, G: FnOnce(HashSet<SocketAddr>) -> Plan<'a>> PlanWithFallback<'a, G> {
+    pub fn new(plan: Plan<'a>, fallback_generator: G) -> Self {
+        Self {
+            source: PlanSource::Primary(TrackedPlan::new(plan)),
+            fallback_generator: Some(fallback_generator),
+        }
+    }
+}
+
+impl<'a, G: FnOnce(HashSet<SocketAddr>) -> Plan<'a>> Iterator for PlanWithFallback<'a, G> {
+    type Item = Arc<Node>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match &mut self.source {
+            PlanSource::Primary(plan) => {
+                let node = plan.next();
```

Comment:
The `Iterator` interface doesn't guarantee that after `.next()` returns `None` it will keep returning `None`. Ref: [link](https://doc.rust-lang.org/std/iter/trait.Iterator.html#tymethod.next)

Reply:
I don't think this code uses such assumption. When `node` is `None` `self.source` switches to `PlanSource::Fallback` causing `next()` method not to be called anymore on the primary plan.